File: RuyiSDKCpp/RuyiNet/Service/RuyiNetLobbyService.cpp
#include "RuyiNetLobbyService.h"

namespace Ruyi
{
	void RuyiNetLobbyService::CreateLobby(int index, int maxSlots, RuyiNetLobbyType lobbyType)
	{
		CreateLobby(index, maxSlots, lobbyType, "{}");
	}

	void RuyiNetLobbyService::CreateLobby(int index, int maxSlots, RuyiNetLobbyType lobbyType, std::string customAttributes)
	{
		nlohmann::json payloadJson;

		payloadJson["appId"] = mClient->GetAppId();
		payloadJson["maxSlots"] = maxSlots;
		payloadJson["ranked"] = (lobbyType == RuyiNetLobbyType::RANKED);
		payloadJson["customAttributes"] = customAttributes;

		std::string responseStr = RunParentScript(index, "Lobby_Create", payloadJson);
		nlohmann::json retJson = nlohmann::json::parse(responseStr);
		RuyiNetLobbyResponse response;
		response.parseJson(retJson);

		InitLobby(response.data.response);
	}

	void RuyiNetLobbyService::CloseLobby(int index, std::string lobbyId, RuyiNetResponse& response)
	{
		DeleteLobby();

		nlohmann::json payloadJson;

		payloadJson["applobbyIdId"] = lobbyId;

		std::string responseStr = RunParentScript(index, "Lobby_Close", payloadJson);
		nlohmann::json retJson = nlohmann::json::parse(responseStr);
		response.parseJson(retJson);
	}

	void RuyiNetLobbyService::FindLobbies(int index, int numResults, RuyiNetLobbyType lobbyType, std::list<RuyiNetLobby*>& lobbyList)
	{
		FindLobbies(index, numResults, lobbyType, 0, lobbyList);
	}

	void RuyiNetLobbyService::FindLobbies(int index, int numResults, RuyiNetLobbyType lobbyType, int freeSlots, std::list<RuyiNetLobby*>& lobbyList)
	{
		FindLobbies(index, numResults, lobbyType, freeSlots, "{}", lobbyList);
	}

	void RuyiNetLobbyService::FindLobbies(int index, int numResults, RuyiNetLobbyType lobbyType, int freeSlots, std::string searchCriteria, std::list<RuyiNetLobby*>& lobbyList)
	{
		nlohmann::json payloadJson;

		payloadJson["appId"] = mClient->GetAppId();
		payloadJson["numResults"] = numResults;
		payloadJson["freeSlots"] = freeSlots;
		payloadJson["ranked"] = (lobbyType == RuyiNetLobbyType::RANKED);
		payloadJson["searchCriteria"] = searchCriteria;

		std::string responseStr = RunParentScript(index, "Lobby_Find", payloadJson);
		nlohmann::json retJson = nlohmann::json::parse(responseStr);
		RuyiNetLobbyFindResponse response;
		response.parseJson(retJson);

		std::for_each(response.data.response.results.items.begin(), response.data.response.results.items.end(), [&](RuyiNetResponseGroup& groupResponse) 
		{
			RuyiNetLobby* pLobby = new RuyiNetLobby(groupResponse);
			lobbyList.push_back(pLobby);
		});	
	}

	void RuyiNetLobbyService::JoinLobby(int index, std::string lobbyId)
	{
		nlohmann::json payloadJson;

		payloadJson["lobbyId"] = lobbyId;

		std::string responseStr = RunParentScript(index, "Lobby_Join", payloadJson);
		nlohmann::json retJson = nlohmann::json::parse(responseStr);
		RuyiNetLobbyResponse response;
		response.parseJson(retJson);

		InitLobby(response.data.response);
	}

	void RuyiNetLobbyService::LeaveLobby(int index, std::string lobbyId, RuyiNetResponse& response)
	{
		nlohmann::json payloadJson;

		payloadJson["lobbyId"] = lobbyId;

		std::string responseStr = RunParentScript(index, "Lobby_Leave", payloadJson);
		nlohmann::json retJson = nlohmann::json::parse(responseStr);
		response.parseJson(retJson);
	}

	void RuyiNetLobbyService::StartGame(int index, std::string lobbyId, std::string connectionString, RuyiNetResponse& response)
	{
		nlohmann::json payloadJson;

		payloadJson["lobbyId"] = lobbyId;
		payloadJson["connectionString"] = connectionString;

		std::string responseStr = RunParentScript(index, "Lobby_StartGame", payloadJson);
		nlohmann::json retJson = nlohmann::json::parse(responseStr);
		response.parseJson(retJson);
	}

	void RuyiNetLobbyService::UpdateLobby()
	{
		if (nullptr != mCurrentLobby)
		{
			nlohmann::json payloadJson;

			payloadJson["lobbyId"] = mCurrentLobby->GetLobbyId();

			std::string responseStr = RunParentScript(mClient->ActivePlayerIndex(), "Lobby_Update", payloadJson);
			nlohmann::json retJson = nlohmann::json::parse(responseStr);
			RuyiNetLobbyResponse response;
			response.parseJson(retJson);

			RuyiNetLobby* pUpdatedLobby = new RuyiNetLobby(response.data.response);
			if (nullptr != pUpdatedLobby)
			{
				std::list<std::string> newPlayers; //in UpdatedLobby but not in CurrentLobby
				std::list<std::string> leftPlayers; //in currentLobby but not in UpdatedLobby which means they've left

				bool lobbyClosed = (0 != mCurrentLobby->GetState().compare("CLOSED")) && (0 == pUpdatedLobby->GetState().compare("CLOSED"));
				bool lobbyStarted = (0 != mCurrentLobby->GetState().compare("STARTED")) && (0 == pUpdatedLobby->GetState().compare("STARTED"));
				std::list<const std::string*>& updatedMemberIds = (std::list<const std::string*>&)pUpdatedLobby->GetMemberProfileIds();
				std::list<const std::string*>& currentMemberIds = (std::list<const std::string*>&)mCurrentLobby->GetMemberProfileIds();
				std::list<const std::string*>::iterator it_update;
				std::list<const std::string*>::iterator it_current;

				//find new players which are not in currentMemberIds but in updatedMemberIds
				for (it_update = updatedMemberIds.begin(); it_update != updatedMemberIds.end(); ++it_update)
				{
					bool isNew = true;

					for (it_current = currentMemberIds.begin(); it_current != currentMemberIds.end(); ++it_current)
					{
						if (0 == (*it_update)->compare(**it_current)) 
						{
							isNew = false;
							break;
						}
					}

					if (isNew)
					{
						newPlayers.push_back(**it_update);
					}
				}
				
				//find left players which are not in updatedMemberIds but in currentMemberIds
				for (it_current = currentMemberIds.begin(); it_current != currentMemberIds.end(); ++it_current)
				{
					bool isOld = true;

					for (it_update = updatedMemberIds.begin(); it_update != updatedMemberIds.end(); ++it_update)
					{
						if (0 == (*it_current)->compare(**it_update))
						{
							isOld = false;
							break;
						}
					}

					if (isOld)
					{
						leftPlayers.push_back(**it_current);
					}
				}

				for (std::list<std::string>::iterator it = newPlayers.begin(); it != newPlayers.end(); ++it)
				{
				
				}
			}
		}
	}

	void RuyiNetLobbyService::InitLobby(RuyiNetResponseGroup& response)
	{
		DeleteLobby();

		mCurrentLobby = new RuyiNetLobby(response);
	}

	void RuyiNetLobbyService::DeleteLobby()
	{
		if (nullptr != mCurrentLobby)
		{
			delete mCurrentLobby;
			mCurrentLobby = nullptr;
		}
	}
}

File: RuyiSDK/RuyiNet/Service/RuyiNetLobbyService.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Timers;

namespace Ruyi
{
    /// <summary>
    /// Manages lobbies for network games.
    /// </summary>
    public class RuyiNetLobbyService : RuyiNetService
    {
        /// <summary>
        /// Creates a lobby that other players can find and join.
        /// </summary>
        /// <param name="index">The index of user</param>
        /// <param name="maxSlots">The maximum number of players that can join this lobby.</param>
        /// <param name="lobbyType">Whether or not this lobby is for a RANKED MATCH or a PLAYER MATCH.</param>
        /// <param name="callback">The function to call when the task completes.</param>
        public void CreateLobby(int index, int maxSlots, RuyiNetLobbyType lobbyType, Action<RuyiNetLobby> callback)
        {
            CreateLobby(index, maxSlots, lobbyType, "{}", callback);
        }

        /// <summary>
        /// Creates a lobby that other players can find and join.
        /// </summary>
        /// <param name="index">The index of user</param>
        /// <param name="maxSlots">The maximum number of players that can join this lobby.</param>
        /// <param name="lobbyType">Whether or not this lobby is for a RANKED MATCH or a PLAYER MATCH.</param>
        /// <param name="customAttributes">JSON string of custom attributes to attach to this lobby.</param>
        /// <param name="callback">The function to call when the task completes.</param>
        public void CreateLobby(int index, int maxSlots, RuyiNetLobbyType lobbyType,
            string customAttributes, Action<RuyiNetLobby> callback)
        {
            var payload = new RuyiNetLobbyCreateRequest()
            {
                appId = mClient.AppId,
                maxSlots = maxSlots,
                ranked = (lobbyType == RuyiNetLobbyType.RANKED),
                customAttributes = customAttributes
            };

            RunPlatformScript(index, "Lobby_Create", JsonConvert.SerializeObject(payload),
                (RuyiNetLobbyResponse response) =>
                {
                    mCurrentLobby = response;
                    callback(mCurrentLobby);
                    PollLobbyStatus();
                });
        }

        /// <summary>
        /// Closes a lobby and kicks all players.
        /// </summary>
        /// <param name="index">The index of user</param>
        /// <param name="lobbyId">The ID of the lobby to close.</param>
        /// <param name="callback">The function to call when the task completes.</param>
        public void CloseLobby(int index, string lobbyId, RuyiNetTask<RuyiNetResponse>.CallbackType callback)
        {
            mCurrentLobby = null;

            var payload = new RuyiNetLobbyCloseRequest() { lobbyId = lobbyId };
            RunPlatformScript(index, "Lobby_Close", JsonConvert.SerializeObject(payload), callback);
        }

        /// <summary>
        /// Searches for lobbies created by other players.
        /// </summary>
        /// <param name="index">The index of user</param>
        /// <param name="numResults">The maximum number of lobbies to return.</param>
        /// <param name="lobbyType">Whether or not this lobby is for a RANKED MATCH or a PLAYER MATCH.</param>
        /// <param name="callback">The function to call when the task completes.</param>
        public void FindLobbies(int index, int numResults, RuyiNetLobbyType lobbyType,
            Action<RuyiNetLobby[]> callback)
        {
            FindLobbies(index, numResults, lobbyType, 0, callback);
        }

        /// <summary>
        /// Searches for lobbies created by other players.
        /// </summary>
        /// <param name="index">The index of user</param>
        /// <param name="numResults">The maximum number of lobbies to return.</param>
        /// <param name="lobbyType">Whether or not this lobby is for a RANKED MATCH or a PLAYER MATCH.</param>
        /// <param name="freeSlots">The number of free slots needed.</param>
        /// <param name="callback">The function to call when the task completes.</param>
        public void FindLobbies(int index, int numResults, RuyiNetLobbyType lobbyType,
            int freeSlots, Action<RuyiNetLobby[]> callback)
        {
            FindLobbies(index, numResults, lobbyType, freeSlots, "{}", callback);
        }

        /// <summary>
        /// Searches for lobbies created by other players.
        /// </summary>
        /// <param name="index">The index of user</param>
        /// <param name="numResults">The maximum number of lobbies to return.</param>
        /// <param name="lobbyType">Whether or not this lobby is for a RANKED MATCH or a PLAYER MATCH.</param>
        /// <param name="freeSlots">The number of free slots needed.</param>
        /// <param name="searchCriteria">JSON string representing parameters to search for.</param>
        /// <param name="callback">The function to call when the task completes.</param>
        public void FindLobbies(int index, int numResults, RuyiNetLobbyType lobbyType,
            int freeSlots, string searchCriteria, Action<RuyiNetLobby[]> callback)
        {
            var payload = new RuyiNetLobbyFindRequest()
            {
                appId = mClient.AppId,
                numResults = numResults,
                freeSlots = freeSlots,
                ranked = (lobbyType == RuyiNetLobbyType.RANKED),
                searchCriteria = searchCriteria
            };

            RunPlatformScript(index, "Lobby_Find", JsonConvert.SerializeObject(payload),
                (RuyiNetLobbyFindResponse response) =>
                {
                    var results = response.data.response.results;
                    var lobbies = new RuyiNetLobby[results.count];
                    for (int i = 0; i < results.count; ++i)
                    {
                        lobbies[i] = new RuyiNetLobby(results.items[i]);
                    }

                    callback(lobbies);
                });
        }

        /// <summary>
        /// Joins a lobby created by another player.
        /// </summary>
        /// <param name="index">The index of user</param>
        /// <param name="lobbyId">The ID of the lobby to join.</param>
        /// <param name="callback">The function to call when the task completes.</param>
        public void JoinLobby(int index, string lobbyId, Action<RuyiNetLobby> callback)
        {
            var payload = new RuyiNetLobbyJoinRequest() { lobbyId = lobbyId };
            RunPlatformScript(index, "Lobby_Join", JsonConvert.SerializeObject(payload),
                (RuyiNetLobbyResponse response) =>
                {
                    mCurrentLobby = response;
                    callback(mCurrentLobby);
                    PollLobbyStatus();
                });
        }

        /// <summary>
        /// Leaves a lobby.
        /// </summary>
        /// <param name="index">The index of user</param>
        /// <param name="lobbyId">The ID of the lobby to leave.</param>
        /// <param name="callback">The function to call when the task completes.</param>
        public void LeaveLobby(int index, string lobbyId, RuyiNetTask<RuyiNetResponse>.CallbackType callback)
        {
            mCurrentLobby = null;

            var payload = new RuyiNetLobbyLeaveRequest() { lobbyId = lobbyId };
            RunPlatformScript(index, "Lobby_Leave", JsonConvert.SerializeObject(payload), callback);
        }

        /// <summary>
        /// Starts a lobby game
        /// </summary>
        /// <param name="index">The index of user</param>
        /// <param name="lobbyId">The ID of the lobby to start the game for.</param>
        /// <param name="connectionString">A connection string used to connect to the host.</param>
        /// <param name="callback">The function to call when the task completes.</param>
        public void StartGame(int index, string lobbyId, string connectionString, RuyiNetTask<RuyiNetResponse>.CallbackType callback)
        {
            var payload = new RuyiNetLobbyStartGameRequest() { lobbyId = lobbyId, connectionString = connectionString };
            RunPlatformScript(index, "Lobby_StartGame", JsonConvert.SerializeObject(payload), callback);
        }

        /// <summary>
        /// Event handler for when a new player joins the lobby.
        /// </summary>
        /// <param name="profileId">The profile ID of the player joining.</param>
        public delegate void PlayerJoinEventHandler(string profileId);

        /// <summary>
        /// Event handler for when a player leaves the lobby.
        /// </summary>
        /// <param name="profileId">The profile ID of the player leaving.</param>
        public delegate void PlayerLeaveEventHandler(string profileId);

        /// <summary>
        /// Event handler for when a lobby should start a game.
        /// </summary>
        public delegate void StartGameEventHandler();

        /// <summary>
        /// Event handler for when the lobby is closed.
        /// </summary>
        public delegate void LobbyClosedEventHandler();

        /// <summary>
        /// Returns the current lobby information.
        /// </summary>
        public RuyiNetLobby CurrentLobby { get { return mCurrentLobby; } }

        /// <summary>
        /// Event handler for when a new player joins the lobby.
        /// </summary>
        public event PlayerJoinEventHandler OnPlayerJoinLobby;

        /// <summary>
        /// Event handler for when a player leaves the lobby.
        /// </summary>
        public event PlayerLeaveEventHandler OnPlayerLeaveLobby;

        /// <summary>
        /// Event handler for when a lobby should start a game.
        /// </summary>
        public event StartGameEventHandler OnLobbyStartGame;

        /// <summary>
        /// Event handler for when the lobby is closed.
        /// </summary>
        public event LobbyClosedEventHandler OnLobbyClosed;

        internal void Update(Object source, ElapsedEventArgs e)
        {
            if (mCurrentLobby != null)
            {
                var payload = new RuyiNetLobbyJoinRequest() { lobbyId = mCurrentLobby.LobbyId };
                RunPlatformScript(mClient.ActivePlayerIndex, "Lobby_Update", JsonConvert.SerializeObject(payload),
                    (RuyiNetLobbyResponse response) =>
                    {
                        RuyiNetLobby updatedLobby = response;
                        if (updatedLobby != null)
                        {
                            if (mCurrentLobby != null)
                            {
                                IEnumerable<string> newPlayers = null;
                                IEnumerable<string> oldPlayers = null;

                                bool lobbyClosed = mCurrentLobby.State != "CLOSED" &&
                                                   updatedLobby.State == "CLOSED";
                                bool lobbyStarted = mCurrentLobby.State != "STARTED" &&
                                                    updatedLobby.State == "STARTED";

                                if (!updatedLobby.MemberProfileIds.SequenceEqual(mCurrentLobby.MemberProfileIds))
                                {
                                    newPlayers = updatedLobby.MemberProfileIds.Except(mCurrentLobby.MemberProfileIds);
                                    oldPlayers = mCurrentLobby.MemberProfileIds.Except(updatedLobby.MemberProfileIds);
                                }

                                mCurrentLobby = updatedLobby;

                                if (newPlayers != null)
                                {
                                    foreach (var i in newPlayers)
                                    {
                                        Console.Write("New Player: " + i);
                                        OnPlayerJoinLobby(i);
                                    }
                                }

                                if (oldPlayers != null)
                                {
                                    foreach (var i in oldPlayers)
                                    {
                                        Console.Write("Old Player: " + i);
                                        OnPlayerLeaveLobby(i);
                                    }
                                }

                                if (lobbyClosed)
                                {
                                    for (var i = 0; i < mClient.CurrentPlayers.Length; ++i)
                                    {
                                        if (mClient.CurrentPlayers[i] != null)
                                        {
                                            LeaveLobby(i, mCurrentLobby.LobbyId, null);
                                        }
                                    }

                                    OnLobbyClosed();
                                }
                                else if (lobbyStarted)
                                {
                                    OnLobbyStartGame();
                                }
                            }
                            else
                            {
                                mCurrentLobby = updatedLobby;
                            }

                            PollLobbyStatus();
                        }
                    });
            }
        }

        /// <summary>
        /// /// Creates a Ruyi Net Service
        /// /// </summary>
        /// /// <param name="client">The Ruyi Net Client.</param>
        internal RuyiNetLobbyService(RuyiNetClient client)
            : base(client)
        {
            mCurrentLobby = null;
        }

        private void PollLobbyStatus()
        {
            mTimer = new Timer(POLL_FREQUENCY);
            mTimer.Elapsed += Update;
            mTimer.AutoReset = false;
            mTimer.Enabled = true;
        }

        private const double POLL_FREQUENCY = 5000;
        private Timer mTimer;
        private RuyiNetLobby mCurrentLobby;
    }
}
